src/lib.rs

```diff
@@ -415,83 +386,187 @@ fn finalize_fips_approval(
     }
 }
 
-extern "C" fn fn_initialize(_init_args: CK_VOID_PTR) -> CK_RV {
-    let mut slotnum: CK_SLOT_ID = 0;
-    let conf: String;
+struct GlobalConfig {
+    conf: Config,
+}
 
-    if _init_args.is_null() {
-        conf = res_or_ret!(find_conf());
-    } else {
-        let args = _init_args as *const CK_C_INITIALIZE_ARGS;
-        if unsafe { (*args).pReserved.is_null() } {
-            conf = res_or_ret!(find_conf());
-        } else {
-            let c_str =
-                unsafe { CStr::from_ptr((*args).pReserved as *const _) };
-            if let Ok(s) = c_str.to_str() {
-                conf = s.to_string();
-            } else {
-                return CKR_ARGUMENTS_BAD;
-            }
+impl GlobalConfig {
+    fn empty_config() -> GlobalConfig {
+        GlobalConfig {
+            conf: Config::new(),
         }
     }
 
-    let v: Vec<&str> = conf.split(':').collect();
-    if v.len() > 1 {
-        slotnum = match CK_SLOT_ID::from_str(v[1]) {
+    fn default_config() -> GlobalConfig {
+        /* if there is no config file we get an empty config */
+        let filename = match Config::find_conf() {
+            Ok(f) => f,
+            Err(_) => return Self::empty_config(),
+        };
+        /* if the file is not accessible or malformed we set an empty config,
+         * an error will be returned later at fn_initialize() time */
+        let conf = match Config::from_file(&filename) {
             Ok(n) => n,
-            Err(_) => return CKR_ARGUMENTS_BAD,
+            Err(_) => {
+                let mut gc = Self::empty_config();
+                /* backwards compatibility where we used to only specify
+                 * a file, this does not support all older options, just
+                 * the more common one of specifying a .sql file with no
+                 * slot specification. */
+                if filename.ends_with(".sql") {
```

Comment:
At least for opensc, the KRYOPTIC_CONF had both db file and slot number (which matches the init args semantics) so this condition wont hit if I see right with the example from OpenSC:
https://github.com/OpenSC/OpenSC/blob/f7070f4c9f9c6e1cb8d2efe838f73d54c78553da/src/tests/p11test/runtest.sh#L128
But we do not have any automation there yet so I can change it in OpenSC tests.

I see the pkcs11-provider uses just the path to the db though:
https://github.com/latchset/pkcs11-provider/blob/main/tests/kryoptic-init.sh#L30

Reply:
Passing in a slot number that way was always mant to be just for testing, I should have wrapped that syntax in #[cfg(test)] indeed.

You should definitely change opensc to use the db=sqlite:/path/to/file.sql,slot=1 or maybe use a config file.

Speaking of which, one question I had about this PR is whether we should always pass parameters this way or if we should also have a way to just pass in a config file name to read from like: config=/path/to/config/file.conf